base/show.jl

```diff
@@ -174,13 +174,36 @@ function show(io::IO, x::Core.IntrinsicFunction)
     print(io, unsafe_string(name))
 end
 
+show(io::IO, ::Core.BottomType) = print(io, "Union{}")
+
 function show(io::IO, x::Union)
     print(io, "Union")
-    sorted_types = sort!(collect(x.types); by=string)
+    sorted_types = sort!(uniontypes(x); by=string)
     show_comma_array(io, sorted_types, '{', '}')
 end
 
-show(io::IO, x::TypeConstructor) = show(io, x.body)
+function print_without_params(x::ANY)
+    if isa(x,UnionAll)
+        b = unwrap_unionall(x)
+        return isa(b,DataType) && b.name.wrapper === x
+    end
+    return false
+end
+
+function show(io::IO, x::UnionAll)
+    if print_without_params(x)
+        return show(io, unwrap_unionall(x).name)
+    end
+    tvar_env = get(io, :tvar_env, false)
+    if tvar_env !== false && isa(tvar_env, AbstractVector)
+        tvar_env = Any[tvar_env..., x.var]
+    else
+        tvar_env = Any[x.var]
+    end
+    show(IOContext(io, tvar_env = tvar_env), x.body)
```

Comment:
This seems to allocate a lot more garbage than necessary. You can just use `IOContext` as an immutable set directly (like we do for SHOWN_SET):
```julia
show(IOContext(io, :unionall_env => x.var), x.body)
```

and later, in `function show(io::IO, tv::TypeVar)`, you can then also shorten the environment test to:
```julia
in_env = (:unionall_env => tv) in io
```